dftimewolf/lib/collectors/gcp_logging.py

```diff
@@ -42,75 +40,121 @@ def __init__(self,
     super(GCPLogsCollector, self).__init__(state, name=name, critical=critical)
     self._filter_expression = ''
     self._project_name = ''
+    self._backoff = False
+    self._delay = 0
+
+  def OutputFile(self):
+    """Generate an output file name and path"""
+    output_file = tempfile.NamedTemporaryFile(
+        mode='w', delete=False, encoding='utf-8', suffix='.jsonl')
+    output_path = output_file.name
+    self.logger.info('Downloading logs to {0:s}'.format(output_path))
+    return output_file, output_path
+
+  def SetupLoggingClient(self):
+    """Sets up a GCP Logging Client
+    
+    Args:
+      N/A
+    """
+    if self._project_name:
+      return logging.Client(_use_grpc=False,  # type: ignore
+                                        project=self._project_name)
+    else:
+      return logging.Client(_use_grpc=False)  # type: ignore
+  
+  def ListPages(self, logging_client):
+    """Returns pages based on a Cloud Logging filter
+    
+    Args:
+      logging_client: A GCP Cloud Logging client
+    """
+    results = logging_client.list_entries(  # type: ignore
+          order_by=logging.DESCENDING,
+          filter_=self._filter_expression,
+          page_size=1000)
+    return results.pages
+
+  def ProcessPages(self, pages, backoff_multiplier, output_file, output_path):
+    """Iterates through a generator or pages and saves logs to disk.
+    Can optionally perform exponential backoff if query API limits are exceeded.
+    
+    Args:
+      pages (generator): A google cloud logging list_entries generator pages object
+      backoff_multiplier (str): Query delay multiplier if the API quota is met and backoff is enabled
+      output_file (str): Output file name
+      output_path (str): Output file path
+    
+    Returns:
+      output_path (str): Log output path (may have been updated if API quota was exceeded)
+    """
+    while True:
+      try:
+        time.sleep(self._delay)
+        page = next(pages)
+      except google_api_exceptions.TooManyRequests as exception:
+        self.PublishMessage('Hit quota limit requesting GCP logs.')
+        if self._backoff is True:
+          self.PublishMessage('Retrying in 60 seconds with a slower query rate.')
+          self.PublishMessage('Due to the GCP logging API, the query must restart from the beginning')
+          time.sleep(60)
+          if self._delay == 0:
+            self._delay = 1
+          else:
+            self._delay *= backoff_multiplier
+          self.PublishMessage('Setting up new logging client.')
```

Comment:
I'd caution against overuse of `self.PublishMessage` for lines that could instead go to `self.logger.<LEVEL>`. PublishMessage sends lines to the curses display and is meant for advising a user of general progress. This line is more relevant to debugging I feel, and so should be sent to the logger instead. (Note, anything to PublishMessage should also go to the logger.)

Reply:
Agree - Changed to self.logger.debug